libfwupd/fwupd-common.c

```diff
@@ -313,12 +313,31 @@ fwupd_build_user_agent (const gchar *package_name, const gchar *package_version)
 gchar *
 fwupd_build_machine_id (const gchar *salt, GError **error)
 {
+	const gchar *fn = NULL;
 	g_autofree gchar *buf = NULL;
+	g_auto(GStrv) fns = g_new0 (gchar *, 5);
 	g_autoptr(GChecksum) csum = NULL;
 	gsize sz = 0;
 
-	/* this has to exist */
-	if (!g_file_get_contents ("/etc/machine-id", &buf, &sz, error))
+	/* one of these has to exist */
+	fns[0] = g_build_filename (SYSCONFDIR, "machine-id", NULL);
+	fns[1] = g_build_filename (LOCALSTATEDIR, "lib", "dbus", "machine-id", NULL);
+	fns[2] = g_strdup ("/etc/machine-id");
+	fns[3] = g_strdup ("/var/lib/dbus/machine-id");
+	for (guint i = 0; fns[i] != NULL; i++) {
+		if (g_file_test (fns[i], G_FILE_TEST_EXISTS)) {
+			fn = fns[i];
+			break;
+		}
+	}
+	if (fn == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_READ,
+				     "The machine-id is not present");
+		return NULL;
```

Comment:
One thought; what happens in the context of a flatpak?  None of those files will "exist" right?
As this function actually gets used by a frontend, I'm wondering if when one of the GUI frontends that are distributed as flatpaks starts building report support I get the suspicion this is still not going to work for them.

So while touching it, my idea is that if _none_ of those exist, there are two options I see:
1. The "daemon" passes up the machine-id (it should have access)
2. Generate a UUID and store it in a persistent data store that the frontend can access.

Reply:
I would go for 2 if possible.

Reply:
I'd err on 1 -- it's useful to have this using the same communication channel as the requests -- so it can work in something like a flatpak sandbox. ENOTIME today.

Reply:
Why not both? If the daemon doesn't pass the machine-id we just generate our own.

Reply:
I think I'd agree with the benefit of doing both - then if you have a new enough daemon you get the correct value and if you have an old daemon without the channel you at least have something consistent that can report it.

Reply:
I've pushed a patch that adds a daemon property, although I'm not quite sure how to use it from `fwupd_build_history_report_json()` -- I suppose we could deprecate that and have a `fwupd_client_build_devices_report_json()` instead that can use the daemon-provided value. I guess the `fwupd_build_history_report_json()` symbol could be replaced with something that always returns `FALSE` as we only have in-tree users. We probably need to export something better if we want GUI apps like GNOME Firmware uploading reports too.

Reply:
I think that's a good idea to make a new function and deprecate the old one to get it ready for gui apps to use.

Reply:
I've exported it in the 2nd patch in this PR, although it's not used yet. I'm working on a bigger PR that moves the report generation into libfwupd and I'll switch to using the daemon-provided value there.